crates/admin/src/cluster_controller/logs_controller.rs

```diff
@@ -1223,3 +1240,259 @@ impl<'a> scheduler::PartitionProcessorPlacementHints
+
+            self
+        }
+
+        fn add_all(&mut self, nodes: Vec<(NodeId, EnumSet<Role>, StorageState)>) {
+            for (node_id, roles, storage_state) in nodes {
+                self.nodes_config
+                    .upsert_node(nodeset_selection::tests::node(
+                        node_id.id().raw_id(),
+                        roles,
+                        storage_state,
+                    ));
+                match node_id {
+                    NodeId::Plain(node_id) => {
+                        self.observed_state.dead_nodes.insert(node_id);
+                    }
+                    NodeId::Generational(node_id) => {
+                        self.observed_state
+                            .alive_nodes
+                            .insert(node_id.as_plain(), node_id);
+                    }
+                }
+            }
+        }
+
+        fn with_nodes<const N: usize>(
+            mut self,
+            ids: [u32; N],
+            roles: EnumSet<Role>,
+            state: StorageState,
+        ) -> Self {
+            for id in ids {
+                self.with_node(id, roles, state);
+            }
+            self
+        }
+
+        fn with_dedicated_admin_node(self, id: u32) -> Self {
+            self.with_nodes(
+                [id],
+                enum_set!(Role::Admin | Role::MetadataStore),
+                StorageState::Disabled,
+            )
+        }
+
+        fn with_dedicated_admin_nodes<const N: usize>(self, ids: [u32; N]) -> Self {
+            self.with_nodes(
+                ids,
+                enum_set!(Role::Admin | Role::MetadataStore),
+                StorageState::Disabled,
+            )
+        }
+
+        fn with_dedicated_worker_nodes<const N: usize>(self, ids: [u32; N]) -> Self {
+            self.with_nodes(ids, enum_set!(Role::Worker), StorageState::Disabled)
+        }
+
+        fn with_dedicated_log_server_nodes<const N: usize>(self, ids: [u32; N]) -> Self {
+            self.with_nodes(ids, enum_set!(Role::LogServer), StorageState::ReadWrite)
+        }
+
+        fn with_mixed_server_nodes<const N: usize>(self, ids: [u32; N]) -> Self {
+            self.with_nodes(
+                ids,
+                enum_set!(Role::Worker | Role::LogServer),
+                StorageState::ReadWrite,
+            )
+        }
+
+        fn build(self) -> MockNodes {
+            MockNodes {
+                nodes_config: self.nodes_config,
+                observed_state: self.observed_state,
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn bootstrap_and_reconfigure_replicated_loglet() -> googletest::Result<()> {
+        let mut nodes = MockNodes::builder()
+            .with_dedicated_admin_node(0)
+            .with_dedicated_log_server_nodes([1, 2, 3, 6])
+            .with_dedicated_worker_nodes([4, 5])
+            .build();
+
+        nodes.kill_node(6); // this node is in nodes_config but starts out offline
```

Comment:
If I add `nodes.kill_nodes(3)`, then the first nodeset that gets selected will be `[1, 2]`. I think this is a bit problematic because we want to wait until we have enough nodes before configuring the first log segment.

Reply:
It correctly returns `None` in this scenario now!